### ISSUE TITLE
Question about packing in executable

### ISSUE THREAD

**Issue by johnnyontheweb** [xBimTeam/XbimWebUI#328]
I'm looking for packing this new version in node.js and npm in an executable with Electron-like methodology. Is any already tested route for that? Any suggestion?

The packaging code could be a feature request.

In my trials, I used `electron-forge` with no success - I'm not good in js code, but I noticed that I cannot find a main method in which to place calls. Any advice? thanks in advance

**Comment by GeorgDangl** [comment]
This isn't really related to this repository.

You should be able to include the xBim viewer just fine in an Electron application, the packaging is independent of the viewer itself.

**Comment by johnnyontheweb** [comment]
I know that this is not strictly related, but, as non-expert, I'm asking some hints for packaging - standard procedure for Electron is not working with this repo (see previous comment on main).

**Comment by GeorgDangl** [comment]
What issue are you hitting?

**Comment by johnnyontheweb** [comment]
I followed this guide: [https://www.electronjs.org/docs/tutorial/quick-start#package-and-distribute-the-application](url)
Then I run from command line `npx @electron-forge/cli import`

and I get this error `npm ERR! could not determine executable to run`, that can be solved with this command `npm i -g @electron-forge/cli` (see [https://stackoverflow.com/questions/64737348/issue-with-electron-quick-start-guide](url) )

Hence, instead of first command, I run `electron-forge import`, that went fine:
```
√ Checking your system
√ Initializing Git Repository
√ Writing modified package.json file
√ Installing dependencies
√ Writing modified package.json file
√ Fixing .gitignore


We have ATTEMPTED to convert your app to be in a format that electron-forge understands.

Thanks for using "electron-forge"!!!
```

Finally, `npm run make`, that gave me this:
```
> @xbim/viewer@2.0.1 make
> electron-forge make

√ Checking your system
√ Resolving Forge Config
We need to package your application before we can make it
⠋ Preparing to Package Application for arch: x64
An unhandled rejection has occurred inside Forge:
Error: packageJSON.main must be set to a valid entry point for your Electron app
    at _default (C:\Users\Giovanni\Desktop\modelliGEN\xBim\XbimWebUI\node_modules\@electron-forge\core\src\api\package.ts:107:11)
    at _default (C:\Users\Giovanni\Desktop\modelliGEN\xBim\XbimWebUI\node_modules\@electron-forge\core\src\api\make.ts:156:5)
    at C:\Users\Giovanni\Desktop\modelliGEN\xBim\XbimWebUI\node_modules\@electron-forge\cli\src\electron-forge-make.ts:44:5

Electron Forge was terminated. Location:
{}
```

I think that this is happening because no `"main": ` method is defined. In Electron quick start, I found that 

> NOTE: If the main field is omitted, Electron will attempt to load an index.js file from the directory containing package.json.

but in this repo there's no index.js. Do you have any advice on this?

**Comment by GeorgDangl** [comment]
This is really a bit out of scope here, and I've never worked with `electron-forge` before, but just as the note says, you'll need an `index.js` file as your main Electron entry point. There, you'll probably want to create a new Electron window and use that to host your application.

But I'd suggest to follow some Electron hello world example first, to get something up in Electron and worry about the xBim part after you've got that.

**Comment by andyward** [comment]
Surely you have to provide the main method? @xbim/viewer is just a library so has no entry point

As Georg says this is really an electron packaging issue, and I don't think any of us are that familiar with developing electron.  But it should be agnostics of the various JS dependencies.

I would

1. Package up hello world
2. Package up hello world using a very basic npm package. ( Something like https://www.npmjs.com/package/upper-case)
3. Package up hello world with the xbim viewer. Don't worry about canvas/or webgl. Just make sure you can call `Viewer.check()`

Only then look at implementing any functionality in the viewer. @xbim/viewer has a dependency on gl-matrix - but I'd hope that gets pulled in by electron-forge

Once you've been through 1-3 and still have issues maybe share a repo with the code so we can take a look.

**Comment by johnnyontheweb** [comment]
Thanks, this is all clearer now about libraries - I packaged up hello world, I can use a simple npm package but still cannot render vars into html. once I solved I get back to you

**Comment by johnnyontheweb** [comment]
I added `require('@xbim/viewer');` in the [repo ](https://github.com/johnnyontheweb/nodetest) but I get:
```
App threw an error during load
ReferenceError: self is not defined
    at Object.<anonymous> (C:\Users\Giovanni\Desktop\modelliGEN\xBim\hello\node_modules\@xbim\viewer\index.js:1:224)
    at Module._compile (internal/modules/cjs/loader.js:1078:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1108:10)
    at Module.load (internal/modules/cjs/loader.js:935:32)
    at Module._load (internal/modules/cjs/loader.js:776:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12789)
    at Module.require (internal/modules/cjs/loader.js:959:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (C:\Users\Giovanni\Desktop\modelliGEN\xBim\hello\main.js:4:1)
    at Module._compile (internal/modules/cjs/loader.js:1078:30)
```

How to ho further than that? I'm stuck because I cannot include the package. thanks in advance

**Comment by johnnyontheweb** [comment]
Thanks a lot, I have to study this...
Is it beyond the (learning) scope of this issue to ask to share the js file?

**Comment by johnnyontheweb** [comment]
It works!

> Congrats you now have a 200MB viewer 😃 !
It's better than no viewer at all! =D

I made some tests, and I saw that, if <p> section is removed from HTML, no viewer is shown. IMHO the reason for this is related to the text shown in your screenshot up here, in which, inexplicably, the part of the text before the first comma is missing.
Is there any way to debug this? Maybe better outside electron? Does it depends on div we have for viewer?

**Comment by GeorgDangl** [comment]
For development, you can just work on a regular web application and run it in Chrome. If you don't need any Electron interop features via IPC or some browser customization, the development experience should be almost identical.
So for quick development, you can just work in the browser and attach e.g. Visual Studio Code to it or debug directly in the Chrome Developer Tools.

**Comment by andyward** [comment]
I think we cam close now, as this well outside the sphere of the xbim viewer...

**Comment by johnnyontheweb** [comment]
Hello,
just a final consideration: I compared the size of this Electron package (196 MB) with the a folder with nginx and old "Viewer" and "Libs" folders (around 7 MB).

**Comment by andyward** [comment]
That would be because Electron is shipping a full version of Chromium and Node in every package.

**Comment by johnnyontheweb** [comment]
I know, I mean that I still cannot understand other things in this new JS code. For example, how to publish it in a web server... Is there any tutorial or guide to do so? I'm not a js dev, any hint would be much appreciated.

**Comment by andyward** [comment]
Can you raise another issue for this please?

**Comment by johnnyontheweb** [comment]
#332
